### PR TITLE
Dim tiled windows when a non-tiled window (including transient windows) is focused

### LINKED ISSUE #99: Add visual indicator that Scratch mode is on
I experience some biugginess with some windows resizing or moving themselves. I try to put them in the scratch mode to avoid issues. However Currently it can get confusing if scratch is on. A visual indicator would be nice

### PR DESCRIPTION
**Pull request by jtaala**
Resolves #99. 

PaperWM makes it easy to see the selected (tiled) window by giving it a coloured border.  However, the same can't be said for non-tiled windows, i.e. scratch windows, floating windows, and transient windows (a transient window is a floating window that takes focus from the parent - e.g. when you open an extension settings from the `extensions` app).

This PR improves this by dimming the current selected (but not focused) tiled window's border - which makes it easier to tell that a scratch, floating, or transient window is currently focused.

It also intelligently handles transient windows (checks if a windows has transient windows that are focused etc.).

In short, it provides  a visual indicator that makes it easier to tell when non-tiled windows are focused.

### CODE REVIEW

File: tiling.js [hunk 1]
```diff
@@ -2581,6 +2581,82 @@ var Spaces = class Spaces extends Map {
...
+    if (metaWindow.get_transient_for()) {
+        return true;
+    }
+    else {
+        return false;
+    }
+}
+
+/**
+ * Returns true iff a metaWindows has at least one transient window.
```
Comment (Lythenas) [inline tiling.js]:
```suggestion
 * Returns true if a metaWindows has at least one transient window.
```

File: tiling.js [hunk 2]
```diff
@@ -2581,6 +2581,82 @@ var Spaces = class Spaces extends Map {
...
+ * @returns 
+ */
+function hasTransient(metaWindow) {
+    if (!metaWindow) {
+        return false;
+    }
+    let hasTransient = false;
+    metaWindow.foreach_transient(t => {
+        hasTransient = true;
+        return;
```
Comment (Lythenas) [inline tiling.js]: Does this return do anything? Do you maybe need to return false or something to stop iterating?

File: tiling.js (hunk 1 above)
Comment (jtaala) [reply tiling.js]: Oh, that was habit there - "iff" stands for "if and only if".  It usually refers to bidirectional conditional statements, and my mind sometimes does this to imply to the developer, or myself, that this method returns true "only if" it has transient windows... false otherwise (which you might think is implied, but from a strict logic point of view "iff" is intended to convey there's no other situations where this method will return true).

File: tiling.js (hunk 2 above)
Comment (jtaala) [reply tiling.js]: Good catch, given it's a lambda function, the return here acts like a `continue`... which means it does nothing here.

**Comment by Lythenas** [comment]
I'm seeing the following error in the logs. And this also breaks PaperWM.
```
JS ERROR: TypeError: this[index] is undefined
moveDone@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:1016:13
_makeEaseCallback/<@resource:///org/gnome/shell/ui/environment.js:153:13
_easeActor@resource:///org/gnome/shell/ui/environment.js:242:17
init/Clutter.Actor.prototype.ease@resource:///org/gnome/shell/ui/environment.js:365:19
addTween@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/utils.js:461:15
layoutColumnSimple@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:444:29
layout@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:524:26
removeWindow@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:745:14
remove_handler@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:2852:11
dynamic_function_ref/<@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/utils.js:104:33
makeScratch@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/scratch.js:77:16
toggle@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/scratch.js:139:20
dynamic_function_ref/<@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/utils.js:104:33
asKeyHandler/<@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/keybindings.js:271:20
```

Although I'm not 100% sure if it is caused by this extension. But I haven't seen it before.

**Comment by jtaala** [comment]
> Although I'm not 100% sure if it is caused by this extension. But I haven't seen it before.

That's interesting, don't think I've seen that one too.  Can you reproduce it?

**Comment by jtaala** [comment]
> Although I'm not 100% sure if it is caused by this extension. But I haven't seen it before.

Can you send me the commit hash you're on?  (if it's different from this PR branch).  Interestingly, this looks like it's coming from the a keydbind actionHandler.  Oh also confirm what gnome ver you're on (44 or your 42 at work one?)

**Comment by jtaala** [comment]
Can't see anything in this PR that would cause that - this PR basically just dims borders under certain conditions.

Try to reproduce - looks like it was something triggered by a keybind (so it's should of happened when pressing some button?).

**Comment by jtaala** [comment]
From the looks of the stacktrace, looks like it might have been a close window keybind (since `remove_handler` was called - did you (or do you) sometimes use multiple windows in a column?

The line that tripped is is updating a property  on windows - so putting a guard (e.g. to ensure this[index] exists is probably worthwhile).

Let me know what you find.

**Comment by Lythenas** [comment]
I managed to also reproduce it on develop. So I will look into fixing this separately.

**Comment by jtaala** [comment]
> I managed to also reproduce it on develop. So I will look into fixing this separately.

Nice pickup @Lythenas!  Yeah, you prob saw in this PR, I'm guessing you were testing with moving windows to scratch layer.

Good work.

**Comment by jtaala** [comment]
Thanks @Lythenas,

Merged in #547 and updated this branch.  Give this one a test/approve when you have time and will merge this one in too.
